File: src/omnia/data/structure/tuple/Sextuplet.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Sextuplet<T> extends Sextuple<T, T, T, T, T, T>, Tuples.AtMostSextuplet<T> {

  static <T> Sextuplet<T> of(T first, T second, T third, T fourth, T fifth, T sixth) {
    return new ImmutableSextuplet<>(first, second, third, fourth, fifth, sixth);
  }

  @Override
  <R> Sextuplet<R> map(Function<? super T, ? extends R> mapper);

  @Override
  Quintuplet<T> dropFirst();

  @Override
  Quintuplet<T> dropSecond();

  @Override
  Quintuplet<T> dropThird();

  @Override
  Quintuplet<T> dropFourth();

  @Override
  Quintuplet<T> dropFifth();

  @Override
  Quintuplet<T> dropSixth();

  @Override
  Septuplet<T> concat(T object);

  @Override
  Octuplet<T> concat(Couple<T, T> other);

  @Override
  Nonuplet<T> concat(Triple<T, T, T> other);

  @Override
  Decuplet<T> concat(Quadruple<T, T, T, T> other);
}

File: src/omnia/data/structure/tuple/Sextuplet.kt
package omnia.data.structure.tuple

import omnia.data.structure.tuple.Tuples.AtMostSextuplet

interface Sextuplet<T : Any> : Sextuple<T, T, T, T, T, T>, AtMostSextuplet<T> {

  override fun <R : Any> map(mapper: (T) -> R): Sextuplet<R>
  override fun dropFirst(): Quintuplet<T>
  override fun dropSecond(): Quintuplet<T>
  override fun dropThird(): Quintuplet<T>
  override fun dropFourth(): Quintuplet<T>
  override fun dropFifth(): Quintuplet<T>
  override fun dropSixth(): Quintuplet<T>
  override fun concat(`object`: T): Septuplet<T>
  override fun concat(other: Couple<T, T>): Octuplet<T>
  override fun concat(other: Triple<T, T, T>): Nonuplet<T>
  override fun concat(other: Quadruple<T, T, T, T>): Decuplet<T>

  companion object {

    fun <T : Any> of(first: T, second: T, third: T, fourth: T, fifth: T, sixth: T): Sextuplet<T> {
      return ImmutableSextuplet(first, second, third, fourth, fifth, sixth)
    }
  }
}